sw/device/lib/testing/dif_uart_test.cc

```diff
@@ -0,0 +1,819 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/testing/mock_mmio.h"
+
+#include "gtest/gtest.h"
+#include "sw/device/lib/base/mmio.h"
+
+extern "C" {
+#include "sw/device/lib/dif/dif_uart.h"
+#include "uart_regs.h"  // Generated.
+}  // extern "C"
+
+namespace {
+  class DifUartTest : public testing::Test, public mock_mmio::MmioTest {
+ protected:
+    dif_uart_t dif_uart_;
+
+    // NCO is calculated as NCO = 2^20 * fbaud / fpclk, so the following values
+    // should result in NCO of 1.
+    dif_uart_config_t dif_uart_config_ = {
+      1,                  // baudrate
+      1048576,            // clk_freq_hz
+      kDifUartDisable,    // parity_enable
+      kDifUartParityEven, // parity
+    };
+};
+
+class DifUartTestUartInitialised : public DifUartTest {
+ public:
+  /**
+   * Initialises the UART state structure without calling the dif_uart_init()
+   * routine.
+   */
+  DifUartTestUartInitialised() {
+    dif_uart_.base_addr = dev().region();
+  };
+};
+
+class DifUartTestBytesSendReceive : public DifUartTestUartInitialised {
+ protected:
+  const uint8_t bytes_send_data_[32] = {
+    'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'b', 'b', 'b',
+    'b', 'b', 'b', 'b', 'b', 'c', 'c', 'c', 'c', 'c', 'c',
+    'c', 'c', 'd', 'd', 'd', 'd', 'd', 'd', 'd', 'd',
+  };
+
+  uint8_t bytes_receive_data_[32] = {0};
+};
+
+class DifUartTestUartInterrupt : public DifUartTestUartInitialised {
+ protected:
+  std::map<dif_uart_interrupt_t, uint32_t> uart_irq_reg_map_ = {
```

Comment:
This should also be a constexpr, but the type should be `std::array<std::pair<...>, N>`.

That said, at that point you should forgo `std::pair` entirely and define a separate struct, so that users can write meaningful field names (like `.irq_type` and `.mask`) instead of first and second. In general, avoid pair and tuple unless the STL forces you to use them.